File: 07_Topological_sorting/Level5/3665.py
# 최종 순위
# https://www.acmicpc.net/problem/3665
# 힌트
# 1. 위상정렬을 이용해 정렬 후 순위를 출력 한다.
import sys
from collections import deque
MAXNUM = 501

def topological_sort():
    q = deque()
    for i in range(1, N + 1):
        if indegree[i] == 0:
            q.append(i)

    while q:
        if len(q)>1:
            return 0
        cur = q.popleft()

        answer.append(team[cur])
        if len(answer) == N:
            return 1

        for i in range(1, N + 1):
            if not adj[cur][i]:
                continue
            adj[cur][i] = False
            indegree[i] -= 1
            if indegree[i] == 0:
                q.append(i)
    return -1

if __name__ == "__main__":
    T = int(input())

    for _ in range(T):
        N = int(input())

        answer = []
        team = [0] * MAXNUM
        order = [0] * MAXNUM
        indegree = [0] * MAXNUM
        adj = [[False] * MAXNUM for _ in range(MAXNUM)]

        team = [0] + list(map(int, sys.stdin.readline().split()))
        for i in range(1, N + 1):
            order[team[i]] = i

        for i in range(1, N):
            for j in range(i+1, N + 1):
                adj[i][j] = True
                indegree[j] += 1

        M = int(input())
        for j in range(M):
            f, s = map(int, sys.stdin.readline().split())
            if order[f] > order[s]:
                f, s = s, f
            node_f, node_s = order[f], order[s]

            adj[node_f][node_s] = False
            indegree[node_s] -= 1

            adj[node_s][node_f] = True
            indegree[node_f] += 1

        result = topological_sort()

        if result == -1:
            print("IMPOSSIBLE")
        elif result == 0:
            print("?")
        else:
            for j in range(len(answer)):
                print(answer[j], end=' ')
            print()

File: 07_Topological_sorting/Level5/3665.java
// Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
// 최종 순위
// https://www.acmicpc.net/problem/3665
// 힌트
// 1. 위상정렬을 이용해 정렬 후 순위를 출력 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static final int MAXNUM = 501;
	static int[] team = new int[MAXNUM];
	static int[] order = new int[MAXNUM];
	static int[] indegree = new int[MAXNUM];
	
	static ArrayList<Integer> answer = new ArrayList<>();
	static boolean[][] adj = new boolean[MAXNUM][MAXNUM];
	static int n, m;
	static BufferedReader br;
	
	static int topologicalSort()
	{
	    Queue<Integer> q = new LinkedList<>();
	    for (int i = 1; i <= n; i++)
	    {
	        if (indegree[i] == 0)
	            q.add(i);
	    }

	    while (q.size()>0)
	    {
	        if (q.size() > 1)
	            return 0; //불확실한 순위
	        int cur = q.poll();

	        answer.add(team[cur]);
	        if (answer.size() == n)
	            return 1; //올바른 순위

	        for (int i = 1; i <= n; i++)
	        {
	            if (!adj[cur][i])
	                continue;
	            adj[cur][i] = false;
	            indegree[i]--;
	            if (indegree[i] == 0)
	                q.add(i);
	        }
	    }
	    return -1; //순위를 정할 수 없음
	}

	static void init() throws IOException
	{
	    answer.clear();
	    Arrays.fill(team, 0);
	    Arrays.fill(order, 0);
	    
	    Arrays.fill(indegree, 0);
	    adj = new boolean[MAXNUM][MAXNUM];
	    
		StringTokenizer st = new StringTokenizer(br.readLine());

	    for (int i = 1; i <= n; i++)
	    {
	        team[i] = Integer.parseInt(st.nextToken());
	        order[team[i]] = i;
	    }

	    for (int i = 1; i < n; i++)
	    {
	        for (int j = i + 1; j <= n; j++)
	        {
	            adj[i][j] = true;
	            indegree[j] += 1;
	        }
	    }
	}

	public static void main(String[] args) throws IOException, NumberFormatException {
		br = new BufferedReader(new InputStreamReader(System.in));
		
		int T = Integer.parseInt(br.readLine());
		
		while (T-- > 0) {
			n = Integer.parseInt(br.readLine());
			init();
			m = Integer.parseInt(br.readLine());
			
	        for (int j = 0; j < m; j++)
	        {
	        	StringTokenizer st = new StringTokenizer(br.readLine());
	        	int f = Integer.parseInt(st.nextToken());
	        	int s = Integer.parseInt(st.nextToken());
	            if (order[f] > order[s]) {
	            	int tmp = f;
	            	f = s;
	            	s = tmp;
	            }
	                
	            int node_f = order[f], node_s = order[s];

	            adj[node_f][node_s] = false;
	            indegree[node_s]--;

	            adj[node_s][node_f] = true;
	            indegree[node_f] += 1;
	        }
			
	        int result = topologicalSort();
	        if (result == -1)
	            System.out.println("IMPOSSIBLE");
	        else if (result == 0)
	            System.out.println("?");
	        else
	        {
	        	for (int j = 0; j < answer.size(); j++)
	            	System.out.print(answer.get(j)+ " ");
	        	System.out.println();
	        }	
		}		
	}
}
